components/brave_wallet/browser/bitcoin/bitcoin_test_utils.cc

```diff
@@ -0,0 +1,203 @@
+    return std::string(parts[1]);
+  }
+
+  return absl::nullopt;
+}
+
+}  // namespace
+
+BitcoinTestRpcServer::BitcoinTestRpcServer(KeyringService* keyring_service,
+                                           PrefService* prefs)
+    : keyring_service_(keyring_service), prefs_(prefs) {
+  shared_url_loader_factory_ =
+      base::MakeRefCounted<network::WeakWrapperSharedURLLoaderFactory>(
+          &url_loader_factory_);
+
+  url_loader_factory_.SetInterceptor(base::BindRepeating(
+      &BitcoinTestRpcServer::RequestInterceptor, base::Unretained(this)));
+}
+BitcoinTestRpcServer::~BitcoinTestRpcServer() = default;
+
+scoped_refptr<network::SharedURLLoaderFactory>
+BitcoinTestRpcServer::GetURLLoaderFactory() {
+  return shared_url_loader_factory_;
+}
+
+bitcoin_rpc::AddressStats BitcoinTestRpcServer::EmptyAddressStats(
+    const std::string& address) {
+  bitcoin_rpc::AddressStats stats;
+  stats.address = address;
+  stats.chain_stats.tx_count = "0";
+  stats.chain_stats.funded_txo_sum = "0";
+  stats.chain_stats.spent_txo_sum = "0";
+  stats.mempool_stats.tx_count = "0";
+  stats.mempool_stats.funded_txo_sum = "0";
+  stats.mempool_stats.spent_txo_sum = "0";
+  return stats;
+}
+
+bitcoin_rpc::AddressStats BitcoinTestRpcServer::TransactedAddressStats(
+    const std::string& address) {
+  bitcoin_rpc::AddressStats stats = EmptyAddressStats(address);
+  stats.chain_stats.tx_count = "1";
+  return stats;
+}
+
+void BitcoinTestRpcServer::RequestInterceptor(
+    const network::ResourceRequest& request) {
+  url_loader_factory_.ClearResponses();
+
+  if (request.method == net::HttpRequestHeaders::kPostMethod &&
+      request.url.path_piece() == "/tx") {
+    auto request_string(request.request_body->elements()
+                            ->at(0)
+                            .As<network::DataElementBytes>()
+                            .AsStringPiece());
+    captured_raw_tx_ = request_string;
+    url_loader_factory_.AddResponse(request.url.spec(), kMockBtcTxid3);
+    return;
+  }
+
+  if (request.url.path_piece() == "/blocks/tip/height") {
+    url_loader_factory_.AddResponse(request.url.spec(),
+                                    base::ToString(mainnet_height_));
+    return;
+  }
+
+  if (auto address = IsAddressStatsRequest(request)) {
+    if (address_stats_map_.contains(*address)) {
+      url_loader_factory_.AddResponse(
+          request.url.spec(),
+          base::ToString(address_stats_map_[*address].ToValue()));
+      return;
+    }
+
+    auto addresses = keyring_service_->GetBitcoinAddresses(account_id_);
+    for (const auto& item : *addresses) {
+      if (item->address_string == *address) {
+        url_loader_factory_.AddResponse(
+            request.url.spec(),
+            base::ToString(TransactedAddressStats(*address).ToValue()));
+        return;
+      }
+    }
+
+    url_loader_factory_.AddResponse(
+        request.url.spec(),
+        base::ToString(EmptyAddressStats(*address).ToValue()));
+    return;
+  }
```

Comment:
@fmarier is it possible in semgrep rules to detect presence of `auto` in `if (auto address = IsAddressStatsRequest(request))` to avoid false positives?

Reply:
@supermassive That's a good idea.

https://github.com/brave/security-action/issues/391